Light Curve File Collection Functionality

tfetherolf opened the pull request:
Most of the changes are related to a new ability to use a collection of light curve files as an alternative to inputting a collection of target pixel files. The results are comparable to using the target pixel files, but the target pixel file input is still preferred. Other modifications include an added input for specific periods to search in the periodogram and the model used to correct the raw lightcurve is available as an output.

Light Curve Files:
![TIC288735205_SIP_lcf](https://user-images.githubusercontent.com/30506178/117527128-747cad00-af7e-11eb-8e27-f28fbaa05826.png)

Target Pixel Files:
![TIC288735205_SIP_tpf](https://user-images.githubusercontent.com/30506178/117527123-6cbd0880-af7e-11eb-900f-401f3a8f023d.png)

tfetherolf committed:
Light Curve File Collection Alternative

Added functionality to alternatively input a collection of Light Curve Files instead of Target Pixel Files. The target flux and background light are based on the "sap_flux" and "flux_bkg" columns in the light curve files.

tfetherolf committed:
Optional Input for Searching Specific Periods

Adds an optional input to define specific periods to search. The default is None, but defining this input will cause min_period, max_period, and nperiods to be ignored.

tfetherolf committed:
Model Added to Output

The model by itself has been included as one of the dictionary outputs.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -56,10 +57,12 @@ def SIP(
-    tpfs : lightkurve.TargetPixelFileCollection
-        A collection of target pixel files from the TESS mission. This can be
-        generated using lightkurve's search functions, for example:
+    tpfs : lightkurve.TargetPixelFileCollection or lk.collections.LightCurveCollection
```
Let's change this to a different variable name, e.g. `input` or `data`

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -80,6 +83,10 @@ def SIP(
         When True, will run SFF detrending.
     sff_kwargs : dict
         Dictionary of SFF key words to pass. See lightkurve's SFFCorrector.
+    custom_periods : None or numpy.ndarray
```
I think this is called something else in `astropy.lombscargle`, let's use the same naming convention

tfetherolf replied:
The `astropy.lombscargle` [documentation](https://docs.astropy.org/en/stable/api/astropy.timeseries.LombScargle.html#astropy.timeseries.LombScargle.power) just uses `frequency`, but I changed the variable name to `given_periods` and slightly updated the description to be more consistent with their language.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
-        tpfs_uncorr = [
-            (tpf + np.nan_to_num(tpf.flux_bkg))[
-                np.isfinite(np.nansum(tpf.flux_bkg, axis=(1, 2)))
+    if ((type(tpfs) is lk.collections.TargetPixelFileCollection) or 
```
`isinstance(tpfs)` is recommended over `type(tpfs)`

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+    if ((type(tpfs) is lk.collections.TargetPixelFileCollection) or 
+            (type(tpfs) is lk.targetpixelfile.TessTargetPixelFile)):
+        # Get the un-background subtracted data
+        print("Target Pixel File Input")
```
We don't want print statements in the code, because they're hard to quiet for a user. If you want a log message, we can using the `logging` package. In this case, I don't think it's useful

tfetherolf replied:
I forgot to delete these lines when I was testing the code. :)

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+        print("Target Pixel File Input")
+        if hasattr(tpfs[0], "flux_bkg"):
+            tpfs_uncorr = [
+                (tpf + np.nan_to_num(tpf.flux_bkg.value))[
```
Nice work dealing with the new lightkurve units

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
-                tpf.to_lightcurve(aperture_mask=aper)
-                for tpf, aper in zip(tpfs_uncorr, apers)
-            ]
+        apers = [
```
I think some of these sections have perhaps not been linted quite right, make sure to run `flake8` on these so they lint properly?

tfetherolf replied:
I checked linting using `flake8` and updated accordingly.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+        )
+        
+    # Check if tpfs is a lightcurve file rather than a target pixel file
+    elif ((type(tpfs) is lk.lightcurve.TessLightCurve) or 
```
Same as above, `isinstance()` recommended

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+        
+    # Check if tpfs is a lightcurve file rather than a target pixel file
+    elif ((type(tpfs) is lk.lightcurve.TessLightCurve) or 
+            (type(tpfs) is lk.collections.LightCurveCollection)):
```
You probably want to check this is a `LightCurveCollection` of `TessLightCurve`'s

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+    # Check if tpfs is a lightcurve file rather than a target pixel file
+    elif ((type(tpfs) is lk.lightcurve.TessLightCurve) or 
+            (type(tpfs) is lk.collections.LightCurveCollection)):
+        print("Lightcurve File Input")
```
Same here, print statements aren't good

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+    elif ((type(tpfs) is lk.lightcurve.TessLightCurve) or 
+            (type(tpfs) is lk.collections.LightCurveCollection)):
+        print("Lightcurve File Input")
+        for lcf in tpfs:
```
This is hard to read. `tpfs` in this case is -not- an instance of `TargetPixelFileCollection`. We should call it something else, like `input` or `data` so that it's easy to read.

christinahedges replied:
I feel that this may break if you pass in a `LightCurve` instance. 

If this is a `LightCurveCollection` then this will iterate through each light curve. If this is a `LightCurve`, this will iterate through each element in the light curve. Make sure there is an error catch above this to force the input data to be some sort of `Collection`.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+            (type(tpfs) is lk.collections.LightCurveCollection)):
+        print("Lightcurve File Input")
+        for lcf in tpfs:
+            lcf.sap_flux.value[~np.isfinite(lcf.sap_flux.value)] = np.nanmedian(lcf.sap_flux.value)
```
Why are w not dropping nans using `lcf.remove_nans()` here? Filling them with median values may not be advisable.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+            lcf.sap_flux.value[~np.isfinite(lcf.sap_flux.value)] = np.nanmedian(lcf.sap_flux.value)
+            lcf.sap_bkg.value[~np.isfinite(lcf.sap_bkg.value)] = np.nanmedian(lcf.sap_bkg.value)
+            lcf.flux = lcf.sap_flux
+            lcf.flux_err = lcf.sap_flux_err
```
I don't like this new part of lightkurve, but I think this is the acceptable fix for now.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+            lcf.sap_bkg.value[~np.isfinite(lcf.sap_bkg.value)] = np.nanmedian(lcf.sap_bkg.value)
+            lcf.flux = lcf.sap_flux
+            lcf.flux_err = lcf.sap_flux_err
+        tpfs_uncorr = [(lcf + np.nan_to_num(lcf.sap_bkg))[np.isfinite(lcf.sap_bkg)] for lcf in tpfs]
```
This needs a better variable name too, it's not a tpf, and it is confusing to read.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+        tpfs_uncorr = [(lcf + np.nan_to_num(lcf.sap_bkg))[np.isfinite(lcf.sap_bkg)] for lcf in tpfs]
+
+        lc = lk.LightCurveCollection(tpfs_uncorr).stitch(lambda x:x).normalize()
+        lc.flux_err.value[~np.isfinite(lc.flux_err.value)] = np.nanmedian(lc.flux_err.value)
```
Is this line redundant?

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
-            .to_sparse()
-            for tpf, bkg_aper in zip(tpfs_uncorr, bkg_apers)
-        ]
+        if ((type(tpfs) is lk.collections.TargetPixelFileCollection) or 
```
Why is this down here in another `if` statement, I believe you could add this to the end of the previous `if` statement which checks if the input is TPFs.

christinahedges commented on src/tess_sip/tess_sip.py:
```diff
@@ -92,65 +99,101 @@ def SIP(
+        elif ((type(tpfs) is lk.lightcurve.TessLightCurve) or 
+                (type(tpfs) is lk.collections.LightCurveCollection):
+            bkgs = [
+                lk.correctors.DesignMatrix(lcf.sap_bkg.value, name="bkg")
```
Why is there a PCA here? I believe `lcf.sap_bkg` is a 1D object, so there is no point in doing pca to it, in fact I'm surprised that doesn't cause an error?

christinahedges reviewed (commented):
Hi @tfetherolf, thanks for the pull request! This is great functionality we definitely need! I have give you some quick comments in your PR which will help improve the readability a little, but overall the functionality is great. If this is a work in progress, feel free to add [WIP] in the title and I can give you another review whenever you feel ready for me to take another look at it. Otherwise, if you'd like to address these comments, we can start the process of getting this merged in.

To make this a really great PR, I think `tess-sip` might need some basic tests, just to show that it imports, and can use TPFs and LCFs. If you like, you can feel free to take a crack at putting some tests in with your PR. If you've not done testing before, I can put in some basic tests and show you where to add some for your PR.

Let me know what you think, great work on the pull request!

christinahedges commented:
Hi @tfetherolf! Is this PR still being worked on?

tfetherolf commented:
@christinahedges So sorry, this fell off my radar somehow! I will get back to this over the next week or two. In the meantime, I updated the title with [WIP] as suggested.

icweaver commented:
Hi @tfetherolf, thanks so much for this PR! Being able to use light curve files directly and the additional methods that they support has been a huge help with my current project =]

tfetherolf committed:
Merge branch 'master' into master

tfetherolf committed:
Replaced `tpfs` variable name with `data`

The variables `tpfs` and `tpfs_uncorr` have been changed to `data` and `data_uncorr`, respectively, in order to be more general.

tfetherolf committed:
Changed `custom_periods` variable to `given_periods`

tfetherolf committed:
Added check for `data` input type

An initial step has been added to check that the input `data` is either a `lk.TargetPixelFileCollection` with all items being `lk.TessTargetPixelFile` objects, or the `data` is a `lk.LightCurveCollection` with all items being `lk.TessLightCurve` objects. All type checks have also been consolidated to the setup section of the code.

tfetherolf committed:
Removed nans and redundancies

tfetherolf committed:
corrected linting using `flake8`

tfetherolf commented:
Sorry it took so long to get to your feedback, @christinahedges!! I'd appreciate some help integrating basic tests. The hardest part of this for me is figuring out how GitHub works...

christinahedges commented:
@tfetherolf, thanks for this PR! I think this is quite close to something workable, but there are a few things I'd like to change (e.g. one keyword name, some of the structure) and I'd like add tests. Rather than going back and forth on this, I'm going to merge this PR, and then add in tests in a second PR. 

Thanks for all your hard work! If you want to see how the tests get added in, check out #6!

christinahedges commented:
@tfetherolf Just to add quickly, the functionality you added does calculate a background periodogram, but that isn't valid in this case. The `sap_bkg` can't be used to calculate a background periodogram in the same was as pixels in the background of target pixel files, so I'm taking that functionality out.

icweaver commented:
Thanks all for your work on this! @christinahedges regarding the background periodogram functionality you mentioned, is there another approach that we should follow, similarly to the discussion in #7?
